File: src/test/java/no/nav/sosialhjelp/soknad/web/rest/ressurser/utdanning/UtdanningRessursTest.java
package no.nav.sosialhjelp.soknad.web.rest.ressurser.utdanning;

import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.utdanning.JsonUtdanning;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository;
import no.nav.sosialhjelp.soknad.business.service.TextService;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.model.exception.AuthorizationException;
import no.nav.sosialhjelp.soknad.domain.model.oidc.StaticSubjectHandlerService;
import no.nav.sosialhjelp.soknad.domain.model.oidc.SubjectHandler;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.utdanning.UtdanningRessurs.UtdanningFrontend;
import no.nav.sosialhjelp.soknad.web.sikkerhet.Tilgangskontroll;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static no.nav.sosialhjelp.soknad.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class UtdanningRessursTest {

    private static final String EIER = "123456789101";
    private static final String BEHANDLINGSID = "123";

    @Mock
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Mock
    private Tilgangskontroll tilgangskontroll;

    @Mock
    private TextService textService;

    @InjectMocks
    private UtdanningRessurs utdanningRessurs;

    @Before
    public void setUp() {
        System.setProperty("environment.name", "test");
        SubjectHandler.setSubjectHandlerService(new StaticSubjectHandlerService());
    }

    @After
    public void tearDown() {
        SubjectHandler.resetOidcSubjectHandlerService();
        System.clearProperty("environment.name");
    }

    @Test
    public void getUtdanningSkalReturnereUtdanningUtenErStudentOgStudentgrad(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(null, null));

        final UtdanningFrontend utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID);
        
        assertThat(utdanningFrontend.erStudent).isNull();
        assertThat(utdanningFrontend.studengradErHeltid).isNull();
    }

    @Test
    public void getUtdanningSkalReturnereUtdanningMedErIkkeStudent(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(Boolean.FALSE, null));

        final UtdanningFrontend utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID);
        
        assertThat(utdanningFrontend.erStudent).isEqualTo(Boolean.FALSE);
        assertThat(utdanningFrontend.studengradErHeltid).isNull();
    }

    @Test
    public void getUtdanningSkalReturnereUtdanningMedErStudent(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(Boolean.TRUE, null));

        final UtdanningFrontend utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID);
        
        assertThat(utdanningFrontend.erStudent).isEqualTo(Boolean.TRUE);
        assertThat(utdanningFrontend.studengradErHeltid).isNull();
    }

    @Test
    public void getUtdanningSkalReturnereUtdanningMedErStudentOgStudentgradHeltid(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(Boolean.TRUE, JsonUtdanning.Studentgrad.HELTID));

        final UtdanningFrontend utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID);
        
        assertThat(utdanningFrontend.erStudent).isEqualTo(Boolean.TRUE);
        assertThat(utdanningFrontend.studengradErHeltid).isEqualTo(Boolean.TRUE);
    }

    @Test
    public void getUtdanningSkalReturnereUtdanningMedErStudentOgStudentgradDeltid(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(Boolean.TRUE, JsonUtdanning.Studentgrad.DELTID));

        final UtdanningFrontend utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID);
        
        assertThat(utdanningFrontend.erStudent).isEqualTo(Boolean.TRUE);
        assertThat(utdanningFrontend.studengradErHeltid).isEqualTo(Boolean.FALSE);
    }

    @Test
    public void putUtdanningSkalSetteUtdanningMedErStudent(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(null, null));

        final UtdanningFrontend utdanningFrontend = new UtdanningFrontend()
                .withErStudent(Boolean.TRUE);
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend);

        final SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        final JsonUtdanning utdanning = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getUtdanning();
        assertThat(utdanning.getKilde()).isEqualTo(JsonKilde.BRUKER);
        assertThat(utdanning.getErStudent()).isEqualTo(Boolean.TRUE);
        assertThat(utdanning.getStudentgrad()).isNull();
    }

    @Test
    public void putUtdanningSkalSetteUtdanningMedErStudentOgStudentgrad(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(null, null));

        final UtdanningFrontend utdanningFrontend = new UtdanningFrontend()
                .withErStudent(Boolean.TRUE)
                .withStudengradErHeltid(Boolean.TRUE);
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend);

        final SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        final JsonUtdanning utdanning = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getUtdanning();
        assertThat(utdanning.getKilde()).isEqualTo(JsonKilde.BRUKER);
        assertThat(utdanning.getErStudent()).isEqualTo(Boolean.TRUE);
        assertThat(utdanning.getStudentgrad()).isEqualTo(JsonUtdanning.Studentgrad.HELTID);
    }

    @Test
    public void putUtdanningSkalSetteUtdanningMedErIkkeStudentOgSletteStudentgrad(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithUtdanning(true, JsonUtdanning.Studentgrad.DELTID));

        final UtdanningFrontend utdanningFrontend = new UtdanningFrontend()
                .withErStudent(Boolean.FALSE)
                .withStudengradErHeltid(Boolean.FALSE);
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend);

        final SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        final JsonUtdanning utdanning = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getUtdanning();
        assertThat(utdanning.getKilde()).isEqualTo(JsonKilde.BRUKER);
        assertThat(utdanning.getErStudent()).isEqualTo(Boolean.FALSE);
        assertThat(utdanning.getStudentgrad()).isNull();
    }

    @Test(expected = AuthorizationException.class)
    public void getUtdanningSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerHarTilgang();

        utdanningRessurs.hentUtdanning(BEHANDLINGSID);

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    @Test(expected = AuthorizationException.class)
    public void putUtdanningSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());

        var utdanningFrontend = new UtdanningFrontend();
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend);

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    private SoknadUnderArbeid catchSoknadUnderArbeidSentToOppdaterSoknadsdata() {
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(soknadUnderArbeidRepository).oppdaterSoknadsdata(argument.capture(), anyString());
        return argument.getValue();
    }

    private SoknadUnderArbeid createJsonInternalSoknadWithUtdanning(Boolean erStudent, JsonUtdanning.Studentgrad studentgrad) {
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getUtdanning()
                .withKilde(JsonKilde.BRUKER)
                .withErStudent(erStudent)
                .withStudentgrad(studentgrad);
        return soknadUnderArbeid;
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/utdanning/UtdanningRessursTest.kt
package no.nav.sosialhjelp.soknad.utdanning

import io.mockk.called
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.runs
import io.mockk.slot
import io.mockk.unmockkObject
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.utdanning.JsonUtdanning.Studentgrad
import no.nav.sosialhjelp.soknad.common.MiljoUtils
import no.nav.sosialhjelp.soknad.common.exceptions.AuthorizationException
import no.nav.sosialhjelp.soknad.common.subjecthandler.StaticSubjectHandlerImpl
import no.nav.sosialhjelp.soknad.common.subjecthandler.SubjectHandlerUtils
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.SoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import no.nav.sosialhjelp.soknad.utdanning.UtdanningRessurs.UtdanningFrontend
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class UtdanningRessursTest {

    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository = mockk()
    private val tilgangskontroll: Tilgangskontroll = mockk()

    private val utdanningRessurs = UtdanningRessurs(tilgangskontroll, soknadUnderArbeidRepository)

    @BeforeEach
    fun setUp() {
        mockkObject(MiljoUtils)
        every { MiljoUtils.isNonProduction() } returns true
        SubjectHandlerUtils.setNewSubjectHandlerImpl(StaticSubjectHandlerImpl())
    }

    @AfterEach
    fun tearDown() {
        SubjectHandlerUtils.resetSubjectHandlerImpl()
        unmockkObject(MiljoUtils)
    }

    @Test
    fun utdanningSkalReturnereUtdanningUtenErStudentOgStudentgrad() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(null, null)

        val utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID)
        assertThat(utdanningFrontend.erStudent).isNull()
        assertThat(utdanningFrontend.studengradErHeltid).isNull()
    }

    @Test
    fun utdanningSkalReturnereUtdanningMedErIkkeStudent() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(java.lang.Boolean.FALSE, null)

        val utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID)
        assertThat(utdanningFrontend.erStudent).isFalse
        assertThat(utdanningFrontend.studengradErHeltid).isNull()
    }

    @Test
    fun utdanningSkalReturnereUtdanningMedErStudent() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(java.lang.Boolean.TRUE, null)

        val utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID)
        assertThat(utdanningFrontend.erStudent).isTrue
        assertThat(utdanningFrontend.studengradErHeltid).isNull()
    }

    @Test
    fun utdanningSkalReturnereUtdanningMedErStudentOgStudentgradHeltid() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(java.lang.Boolean.TRUE, Studentgrad.HELTID)

        val utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID)
        assertThat(utdanningFrontend.erStudent).isTrue
        assertThat(utdanningFrontend.studengradErHeltid).isTrue
    }

    @Test
    fun utdanningSkalReturnereUtdanningMedErStudentOgStudentgradDeltid() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(true, Studentgrad.DELTID)

        val utdanningFrontend = utdanningRessurs.hentUtdanning(BEHANDLINGSID)
        assertThat(utdanningFrontend.erStudent).isTrue
        assertThat(utdanningFrontend.studengradErHeltid).isFalse
    }

    @Test
    fun putUtdanningSkalSetteUtdanningMedErStudent() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(null, null)

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val utdanningFrontend = UtdanningFrontend(erStudent = true, studengradErHeltid = null)
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend)

        val soknadUnderArbeid = soknadUnderArbeidSlot.captured
        val utdanning = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.utdanning
        assertThat(utdanning.kilde).isEqualTo(JsonKilde.BRUKER)
        assertThat(utdanning.erStudent).isTrue
        assertThat(utdanning.studentgrad).isNull()
    }

    @Test
    fun putUtdanningSkalSetteUtdanningMedErStudentOgStudentgrad() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(null, null)

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val utdanningFrontend = UtdanningFrontend(erStudent = true, studengradErHeltid = true)
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend)

        val soknadUnderArbeid = soknadUnderArbeidSlot.captured
        val utdanning = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.utdanning
        assertThat(utdanning.kilde).isEqualTo(JsonKilde.BRUKER)
        assertThat(utdanning.erStudent).isTrue
        assertThat(utdanning.studentgrad).isEqualTo(Studentgrad.HELTID)
    }

    @Test
    fun putUtdanningSkalSetteUtdanningMedErIkkeStudentOgSletteStudentgrad() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithUtdanning(true, Studentgrad.DELTID)

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val utdanningFrontend = UtdanningFrontend(erStudent = false, studengradErHeltid = false)
        utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend)

        val soknadUnderArbeid = soknadUnderArbeidSlot.captured
        val utdanning = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.utdanning
        assertThat(utdanning.kilde).isEqualTo(JsonKilde.BRUKER)
        assertThat(utdanning.erStudent).isFalse
        assertThat(utdanning.studentgrad).isNull()
    }

    @Test
    fun utdanningSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { utdanningRessurs.hentUtdanning(BEHANDLINGSID) }

        verify { soknadUnderArbeidRepository wasNot called }
    }

    @Test
    fun putUtdanningSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } throws AuthorizationException("Not for you my friend")

        val utdanningFrontend = UtdanningFrontend(null, null)
        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { utdanningRessurs.updateUtdanning(BEHANDLINGSID, utdanningFrontend) }

        verify { soknadUnderArbeidRepository wasNot called }
    }

    private fun createJsonInternalSoknadWithUtdanning(
        erStudent: Boolean?,
        studentgrad: Studentgrad?
    ): SoknadUnderArbeid {
        val soknadUnderArbeid = SoknadUnderArbeid(
            versjon = 1L,
            behandlingsId = BEHANDLINGSID,
            tilknyttetBehandlingsId = null,
            eier = EIER,
            jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
            status = SoknadUnderArbeidStatus.UNDER_ARBEID,
            opprettetDato = LocalDateTime.now(),
            sistEndretDato = LocalDateTime.now()
        )
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.utdanning
            .withKilde(JsonKilde.BRUKER)
            .withErStudent(erStudent)
            .withStudentgrad(studentgrad)
        return soknadUnderArbeid
    }

    companion object {
        private const val EIER = "12345678910"
        private const val BEHANDLINGSID = "123"
    }
}
